### ISSUE TITLE
MacOS headers are suspiciously incomplete

### ISSUE THREAD

**Issue by romansp** [skpm/sketch-headers#8]
For instance [NSFont.json](https://github.com/skpm/sketch-headers/blob/gh-pages/latest/macos/NSFont.json) contains only 2 methods and a single property. But it should definitely include a whole bunch of methods and properties: https://developer.apple.com/documentation/appkit/nsfont?language=objc

Going to generate and inspect raw headers tomorrow on my dev machine. Will post results later.

**Comment by romansp** [comment]
Well, the reason is that class-dump doesn't generate include all the methods in the header files... Don't know it's happening.

Fortunately, AppKit and Foundation have their headers already included in `/System/Library/Frameworks/<framework>/Headers` directory. We can just parse those additionally to the ones generated by class-dump.

We should probably change the `parseHeaders` to accept `input` array, and iterate over it.
https://github.com/skpm/sketch-headers/blob/fd765dd785d9deb518a7e8576bed0e74739cd883/generate-headers.js#L49-L52

Already tried that change locally. Even though it generates much more, parsing regexes should be revisited once again 😢.

**Comment by mathieudutour** [comment]
oh we should use the `*/Headers` directories directly yes! I think it's worth it, even if we need to change the regexes (maybe need another parser altogether?). We will have the real name of the arguments that way

**Comment by romansp** [comment]
I will add PR tomorrow that includes AppKit and Foundation `*/Headers` to parsing.

Even though we may try using https://github.com/danielmschmidt/objective-c-parse but I don't see support for properties, imports or interfaces/extends there. I will inspect regexes that are used, maybe those are better.

**Comment by romansp** [comment]
For parsing there is an ultimate solution, though it feels like the idea for completely separate project. 

There is ANTLR obj-c grammar available over here https://github.com/antlr/grammars-v4/tree/master/objc. From that grammar it's possible to generate lexer and parser implemented in JavaScript (or C#, Java, Python, C++, Go). And then parse obj-c headers using this generated lexer/parser. Basically parsing result will be an abstract syntax tree (or AST), that can be worked with. Traversing that AST it's possible to generate JSON representation.

Obvious cons are steep learning curve, and maybe a total overkill to `sketch-headers` initial goal.

I worked little bit with ANTLR in C#, it's not THAT hard actually. Static typing saved a lot of time debugging. Don't know what to expect in JavaScript ANTLR.

That's food for thought. Currently hesitating if it even worth this efforts at all.

**Comment by mathieudutour** [comment]
yeah there is also a pegjs grammar for obj-c: https://github.com/okaxaki/objc2swift/blob/master/grammer/objc.pegjs but it does feel like an overkill.

Do you think we don't have a chance at fixing the regex? the headers from `class-dump` are the same all the time so it's mostly about the macOS headers

**Comment by romansp** [comment]
Actually I haven't looked at the regexes yet. It's always possible to fix them, it's just the matter of staying sane 🤣.
